From 00e9cb6f1fadd18e01619d4d40fc69295186206d Mon Sep 17 00:00:00 2001
From: Amalgamator Bench <amalgamator-bench@redhat.com>
Date: Mon, 19 Oct 2026 09:50:24 +0000
Subject: [PATCH 1/2] Port test files for
 crate-fix-replication-from-translog-for-integral (Phase 0.5c)

Ported before fix to enable RED/GREEN checking
---
 .../reference/doc/lucene/SourceParserTest.java | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/server/src/test/java/io/crate/expression/reference/doc/lucene/SourceParserTest.java b/server/src/test/java/io/crate/expression/reference/doc/lucene/SourceParserTest.java
index 15c4236f5..93ef60337 100644
--- a/server/src/test/java/io/crate/expression/reference/doc/lucene/SourceParserTest.java
+++ b/server/src/test/java/io/crate/expression/reference/doc/lucene/SourceParserTest.java
@@ -24,6 +24,7 @@ package io.crate.expression.reference.doc.lucene;
 import static io.crate.testing.TestingHelpers.createReference;
 import static org.assertj.core.api.Assertions.assertThat;
 
+import java.math.BigDecimal;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
@@ -527,4 +528,21 @@ public class SourceParserTest extends ESTestCase {
 
         assertThat(result.get("x")).isNull();
     }
+
+    @Test
+    public void test_ignored_object_with_big_integral_number_is_parsed_as_bigdecimal() throws Exception {
+        SourceParser sourceParser = new SourceParser(Set.of(), UnaryOperator.identity(), true);
+        ObjectType objectType = ObjectType.of(ColumnPolicy.IGNORED).build();
+        sourceParser.register(ColumnIdent.of("_doc", List.of("o")), objectType);
+        Map<String, Object> result = sourceParser.parse(new BytesArray(
+            """
+                {
+                    "o": {
+                        "x": 30000000000000000000000000000000
+                    }
+                }
+            """
+        ));
+        assertThat(Maps.getByPath(result, "o.x")).isInstanceOf(BigDecimal.class);
+    }
 }
-- 
2.53.0


From e0eb4cb8282eda76d0f0f865c3f6a6524f9c1fba Mon Sep 17 00:00:00 2001
From: baur <baurzhansahariev@gmail.com>
Date: Tue, 14 Oct 2025 12:51:12 +0200
Subject: [PATCH 2/2] Fix replication from translog for integral numeric values

(cherry picked from commit 43de35c74ab4d625c289f68a66197c21e923f550)
---
 .../reference/doc/lucene/SourceParser.java     | 18 +++++++++++++++++-
 1 file changed, 17 insertions(+), 1 deletion(-)

diff --git a/server/src/main/java/io/crate/expression/reference/doc/lucene/SourceParser.java b/server/src/main/java/io/crate/expression/reference/doc/lucene/SourceParser.java
index b3f57a89c..d7e77b924 100644
--- a/server/src/main/java/io/crate/expression/reference/doc/lucene/SourceParser.java
+++ b/server/src/main/java/io/crate/expression/reference/doc/lucene/SourceParser.java
@@ -27,6 +27,8 @@ import static org.elasticsearch.common.xcontent.XContentParser.Token.VALUE_NULL;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.UncheckedIOException;
+import java.math.BigDecimal;
+import java.math.BigInteger;
 import java.util.ArrayList;
 import java.util.BitSet;
 import java.util.HashMap;
@@ -288,7 +290,7 @@ public final class SourceParser {
             case START_ARRAY -> parseArray(parser, type, requiredColumns, colPath);
             case START_OBJECT -> parseObject(parser, requiredColumns, colPath, includeUnknown);
             case VALUE_STRING -> isUndefined(type) ? parser.text() : parseByType(parser, type);
-            case VALUE_NUMBER -> isUndefined(type) ? parser.numberValue() : parseByType(parser, type);
+            case VALUE_NUMBER -> isUndefined(type) ? numberValue(parser.numberValue()) : parseByType(parser, type);
             case VALUE_BOOLEAN -> isUndefined(type) ? parser.booleanValue() : parseByType(parser, type);
             case VALUE_EMBEDDED_OBJECT -> isUndefined(type) ? parser.binaryValue() : parseByType(parser, type);
             default -> throw new UnsupportedOperationException("Unsupported token encountered, expected a value, got "
@@ -296,6 +298,20 @@ public final class SourceParser {
         };
     }
 
+    /**
+     * For the undefined case, Number's concrete types is guessed by the parser.
+     * ParserBase.getNumberValue always returns BigInteger for integral big numbers.
+     * We must use BigDecimal for NUMERIC even for integral values, otherwise streaming will be broken.
+     * @param number is a Number returned by the parser.
+     */
+    private static Number numberValue(Number number) {
+        if (number instanceof BigInteger bigInt) {
+            return new BigDecimal(bigInt);
+        }
+        return number;
+    }
+
+
     private static boolean isUndefined(@Nullable DataType<?> type) {
         return type == null || type.id() == DataTypes.UNDEFINED.id();
     }
-- 
2.53.0

